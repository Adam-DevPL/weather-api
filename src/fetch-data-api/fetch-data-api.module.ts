import { HttpModule } from '@nestjs/axios';
import { Module } from '@nestjs/common';
import { FetchDataApiFilter } from 'src/filters/fetch-data-api.filter';
import { FetchDataApiService } from './fetch-data-api.service';
import { APP_FILTER } from '@nestjs/core';

@Module({
  imports: [HttpModule],
  providers: [
    FetchDataApiService,
    {
      provide: APP_FILTER,
      useClass: FetchDataApiFilter,
    },
  ],
  exports: [FetchDataApiService],
})
export class FetchDataApiModule {}
