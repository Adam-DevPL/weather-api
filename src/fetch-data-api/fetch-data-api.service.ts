import { HttpService } from '@nestjs/axios';
import { HttpException, HttpStatus, Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';

import { firstValueFrom } from 'rxjs';
import { LocationType } from 'src/prediction/types/prediction.types';
import {
  FetchDataApiGeoResponse,
  FetchDataApiParams,
} from './types/FetchDataApi.types';

@Injectable()
export class FetchDataApiService {
  private readonly weatherUrl;
  private readonly geoUrl;

  constructor(
    private readonly httpService: HttpService,
    private readonly configService: ConfigService,
  ) {
    this.weatherUrl = this.configService.get<string>('WEATHER_URL');
    this.geoUrl = this.configService.get<string>('GEO_URL');
  }

  public async getDataFromApi(params: FetchDataApiParams) {
    try {
      const { data } = await firstValueFrom(
        this.httpService.get(this.weatherUrl, { params }),
      );
      if (data.error) {
        throw new HttpException(
          { status: HttpStatus.BAD_REQUEST, message: 'Invalid coordinates' },
          HttpStatus.BAD_REQUEST,
        );
      }

      return {
        data,
      };
    } catch ({ response }) {
      throw new HttpException(
        {
          status:
            response.status === 404
              ? HttpStatus.INTERNAL_SERVER_ERROR
              : HttpStatus.BAD_REQUEST,
          error:
            response.status === 404
              ? 'Internal server error'
              : response.message,
        },
        response.status === 404
          ? HttpStatus.INTERNAL_SERVER_ERROR
          : HttpStatus.BAD_REQUEST,
        {
          cause: response,
        },
      );
    }
  }

  public async getGeoLocation(name: string): Promise<FetchDataApiGeoResponse> {
    try {
      const params = {
        name,
        count: 1,
      };

      const { data } = await firstValueFrom(
        this.httpService.get(this.geoUrl, { params }),
      );

      if (!data.hasOwnProperty('results')) {
        throw new HttpException(
          { status: HttpStatus.BAD_REQUEST, message: 'Location not found' },
          HttpStatus.BAD_REQUEST,
        );
      }

      const {
        results: [{ latitude, longitude, country }],
      } = data;

      return {
        latitude,
        longitude,
        location:
          country.toLowerCase() === name.toLowerCase()
            ? LocationType.Country
            : LocationType.City,
      };
    } catch ({ response }) {
      throw new HttpException(
        {
          status:
            response.status === 404
              ? HttpStatus.INTERNAL_SERVER_ERROR
              : HttpStatus.BAD_REQUEST,
          error:
            response.status === 404
              ? 'Internal server error'
              : response.message,
        },
        response.status === 404
          ? HttpStatus.INTERNAL_SERVER_ERROR
          : HttpStatus.BAD_REQUEST,
        {
          cause: response,
        },
      );
    }
  }
}
