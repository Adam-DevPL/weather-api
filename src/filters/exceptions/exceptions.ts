export class LocationException extends Error {
  constructor(readonly message: string) {
    super(message);
    this.name = 'LocationException';
  }
}

export class CoordinatesException extends Error {
  constructor(readonly message: string) {
    super(message);
    this.name = 'CoordinatesException';
  }
}

export class InternalServerException extends Error {
  constructor(readonly message: string) {
    super(message);
    this.name = 'InternalServerException';
  }
}

export class LocationTypeException extends Error {
  constructor(readonly message: string) {
    super(message);
    this.name = 'LocationTypeException';
  }
}
