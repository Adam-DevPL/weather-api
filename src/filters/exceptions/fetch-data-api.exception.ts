export class FetchDataApiException {
  constructor(readonly message: string) {}
}
