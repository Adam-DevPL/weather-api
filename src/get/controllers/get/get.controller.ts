import { Controller, Get, Param } from '@nestjs/common';

@Controller('weather/get')
export class GetController {
  @Get('country/:country')
  getAverageTemperatureForCountry(@Param('country') countryName: string) {
    console.log(countryName);
  }
}
