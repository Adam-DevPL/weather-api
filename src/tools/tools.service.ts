import { Injectable } from '@nestjs/common';

@Injectable()
export class ToolsService {
  public getAverage = (arr: number[]): number => {
    return arr.reduce((acc, item) => acc + item, 0) / arr.length;
  };

  public mostFrequent = (arr: number[]): string =>
    Object.entries(
      arr.reduce((a, v) => {
        a[v] = a[v] ? a[v] + 1 : 1;
        return a;
      }, {}),
    ).reduce((a, v) => (v[1] >= a[1] ? v : a), [null, 0])[0];

  public getWeatherCode = (code: number): string => {
    switch (code) {
      case 0:
        return 'Clear sky';
      case 1:
      case 2:
      case 3:
        return 'Mainly clear, partly cloudy, and overcast';
      case 45:
      case 48:
        return 'Fog and depositing rime fog';
      case 51:
      case 53:
      case 55:
        return 'Drizzle: Light, moderate, and dense intensity';
      case 56:
      case 57:
        return 'Freezing Drizzle: Light and dense intensity';
      case 61:
      case 63:
      case 65:
        return 'Rain: Slight, moderate and heavy intensity';
      case 66:
      case 67:
        return 'Freezing Rain: Light and heavy intensity';
      case 71:
      case 73:
      case 75:
        return 'Snow fall: Slight, moderate, and heavy intensity';
      case 77:
        return 'Snow grains';
      case 80:
      case 81:
      case 82:
        return 'Rain showers: Slight, moderate, and violent';
      case 85:
      case 86:
        return 'Snow showers slight and heavy';
      case 95:
        return 'Thunderstorm: Slight or moderate';
      case 96:
      case 99:
        return 'Thunderstorm with slight and heavy hail';
      default:
        return 'No more weather details';
    }
  };
}
