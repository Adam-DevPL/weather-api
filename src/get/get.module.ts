import { Module } from '@nestjs/common';
import { FetchDataApiModule } from 'src/utils/fetch-data-api/fetch-data-api.module';
import { ToolsModule } from 'src/utils/tools/tools.module';
import { GetController } from './controllers/get/get.controller';
import { GetService } from './services/get/get.service';

@Module({
  imports: [FetchDataApiModule, ToolsModule],
  controllers: [GetController],
  providers: [GetService],
})
export class GetModule {}
