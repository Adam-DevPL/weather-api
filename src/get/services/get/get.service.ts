import { HttpException, HttpStatus, Injectable } from '@nestjs/common';
import {
  GetWeatherResponse,
  GetCountryTempResponse,
  GetCountryParam,
  GetCityParam,
  GetLocationParam,
} from 'src/get/types/get.types';
import { FetchDataApiService } from 'src/utils/fetch-data-api/fetch-data-api.service';
import { FetchDataApiParams } from 'src/utils/fetch-data-api/types/FetchDataApi.types';
import { ToolsService } from 'src/utils/tools/tools.service';

@Injectable()
export class GetService {
  constructor(
    private fetchDataApi: FetchDataApiService,
    private tools: ToolsService,
  ) {}

  public async getCountryAvgTemp({
    country,
  }: GetCountryParam): Promise<GetCountryTempResponse> {
    const { latitude, longitude, location } =
      await this.fetchDataApi.getGeoLocation(country);

    if (location.toLowerCase() !== country.toLowerCase()) {
      throw new HttpException(
        `It is not a country: ${country}`,
        HttpStatus.BAD_REQUEST,
      );
    }

    const fetchDataApiParams: FetchDataApiParams = {
      latitude,
      longitude,
      current_weather: true,
    };

    const {
      data: {
        current_weather: { temperature },
      },
    } = await this.fetchDataApi.getDataFromApi(fetchDataApiParams);

    return {
      country,
      avgTemperatue: temperature,
    };
  }

  public async getCityWeather({
    city,
  }: GetCityParam): Promise<GetWeatherResponse> {
    const { latitude, longitude, location } =
      await this.fetchDataApi.getGeoLocation(city);

    if (location.toLowerCase() === city.toLowerCase()) {
      throw new HttpException(
        `It is not a city: ${city}`,
        HttpStatus.BAD_REQUEST,
      );
    }

    const fetchDataApiParams: FetchDataApiParams = {
      latitude,
      longitude,
      current_weather: true,
    };

    const {
      data: {
        current_weather: { temperature, weathercode },
      },
    } = await this.fetchDataApi.getDataFromApi(fetchDataApiParams);

    return {
      city: city,
      temperature,
      weather: this.tools.getWeatherCode(weathercode),
    };
  }

  public async getLocationWeather({
    lat,
    lon,
  }: GetLocationParam): Promise<GetWeatherResponse> {
    const fetchDataApiParams: FetchDataApiParams = {
      latitude: lat,
      longitude: lon,
      current_weather: true,
    };

    const {
      data: {
        current_weather: { temperature, weathercode },
      },
    } = await this.fetchDataApi.getDataFromApi(fetchDataApiParams);

    return {
      temperature,
      weather: this.tools.getWeatherCode(weathercode),
    };
  }
}
