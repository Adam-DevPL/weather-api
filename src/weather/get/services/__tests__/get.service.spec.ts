import { HttpModule } from '@nestjs/axios';
import { Test, TestingModule } from '@nestjs/testing';
import { FetchDataApiService } from 'src/fetch-data-api/fetch-data-api.service';
import { ToolsService } from 'src/tools/tools.service';
import { GetService } from '../get.service';
import { ConfigService } from '@nestjs/config';
import {
  CoordinatesInput,
  LocationNameInput,
  LocationType,
} from 'src/weather/types/weather.types';
import { LocationException } from 'src/filters/exceptions/exceptions';

describe('GetService', () => {
  let service: GetService;
  let fetchDataApi: FetchDataApiService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [HttpModule],
      providers: [GetService, FetchDataApiService, ToolsService, ConfigService],
    }).compile();

    service = module.get<GetService>(GetService);
    fetchDataApi = module.get<FetchDataApiService>(FetchDataApiService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  describe('getCurrentWeatherWithoutCoordinates', () => {
    it('should return a weather resposne for country', async () => {
      //given
      const locationNameParam: LocationNameInput = {
        locationType: LocationType.COUNTRY,
        locationName: 'Poland',
      };

      jest.spyOn(fetchDataApi, 'getGeoLocation').mockResolvedValue({
        latitude: 52,
        longitude: 20,
        location: LocationType.COUNTRY,
      });

      jest
        .spyOn(service, <any>'getApiData')
        .mockResolvedValue({ avgTemperature: 10, weather: 'Rain' });

      //when
      const result = await service.getCurrentWeatherWithoutCoordinates(
        locationNameParam,
      );

      //then
      expect(result).toMatchSnapshot();
    });

    it('should return a weather resposne for city', async () => {
      //given
      const locationNameParam: LocationNameInput = {
        locationType: LocationType.CITY,
        locationName: 'City',
      };

      jest.spyOn(fetchDataApi, 'getGeoLocation').mockResolvedValue({
        latitude: 52,
        longitude: 20,
        location: LocationType.CITY,
      });

      jest
        .spyOn(service, <any>'getApiData')
        .mockResolvedValue({ avgTemperature: 10, weather: 'Rain' });

      //when
      const result = await service.getCurrentWeatherWithoutCoordinates(
        locationNameParam,
      );

      //then
      expect(result).toMatchSnapshot();
    });

    it('should throw error for invalid parameters, getting City instead of Country', async () => {
      //given
      const locationNameParam: LocationNameInput = {
        locationType: LocationType.COUNTRY,
        locationName: 'City',
      };

      jest.spyOn(fetchDataApi, 'getGeoLocation').mockResolvedValue({
        latitude: 52,
        longitude: 20,
        location: LocationType.CITY,
      });

      await expect(
        service.getCurrentWeatherWithoutCoordinates(locationNameParam),
      ).rejects.toMatchSnapshot();
    });

    it('should throw error for invalid parameters - location name in invalid', async () => {
      //given
      const locationNameParam: LocationNameInput = {
        locationType: LocationType.COUNTRY,
        locationName: 'xxxxxx',
      };

      jest.spyOn(fetchDataApi, 'getGeoLocation').mockImplementation(() => {
        throw new LocationException('Location not found');
      });

      await expect(
        service.getCurrentWeatherWithoutCoordinates(locationNameParam),
      ).rejects.toMatchSnapshot();
    });
  });

  describe('getCurrentWeatherWithCoordinates', () => {
    it('should return weather response for coordinates', async () => {
      const coordinatesInput: CoordinatesInput = {
        lat: 20,
        lon: 20,
      };

      jest
        .spyOn(service, <any>'getApiData')
        .mockResolvedValue({ avgTemperature: 10, weather: 'Rain' });

      //when
      const result = await service.getCurrentWeatherWithCoordinates(
        coordinatesInput,
      );

      //then
      expect(result).toMatchSnapshot();
    });
  });
});
