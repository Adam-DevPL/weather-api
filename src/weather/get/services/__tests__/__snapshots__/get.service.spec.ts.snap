// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetService getCurrentWeatherWithCoordinates should return weather response for coordinates 1`] = `
{
  "avgTemperature": 10,
  "location": {
    "latitude": 20,
    "longitude": 20,
  },
  "weather": "Rain",
}
`;

exports[`GetService getCurrentWeatherWithoutCoordinates should return a weather resposne for city 1`] = `
{
  "avgTemperature": 10,
  "location": "City",
  "weather": "Rain",
}
`;

exports[`GetService getCurrentWeatherWithoutCoordinates should return a weather resposne for country 1`] = `
{
  "avgTemperature": 10,
  "location": "Poland",
  "weather": "Rain",
}
`;

exports[`GetService getCurrentWeatherWithoutCoordinates should throw error for invalid parameters - location name in invalid 1`] = `[LocationException: Location not found]`;

exports[`GetService getCurrentWeatherWithoutCoordinates should throw error for invalid parameters, getting City instead of Country 1`] = `[LocationTypeException: It is not a COUNTRY: City]`;
